File: python-sdk/src/pvium_sdk/services/invoice/endpoints.py
from __future__ import annotations

from typing import Any, Dict, Optional
from urllib.parse import quote

from ...core.client import PviumHttpClient
from ...core.types import RequestOptions


class PviumEndpoints:
    def __init__(self, http: PviumHttpClient):
        self.http = http

    def createInvoice(self, body: Dict[str, Any], options: Optional[RequestOptions] = None) -> Any:
        response = self.http.request("POST", "/v1/invoices", body=body, options=options)
        return self.http.parseResponseBody(response)

    def listInvoices(self, options: Optional[RequestOptions] = None) -> Any:
        response = self.http.request("GET", "/v1/invoices", options=options)
        return self.http.parseResponseBody(response)

    def getInvoiceStatus(self, code: str, options: Optional[RequestOptions] = None) -> Any:
        response = self.http.request(
            "GET",
            f"/v1/invoices/{quote(code, safe='')}/status",
            options=options,
        )
        return self.http.parseResponseBody(response)

    def cancelInvoice(self, invoice_id: str | int, options: Optional[RequestOptions] = None) -> Any:
        response = self.http.request(
            "PATCH",
            f"/v1/invoices/{quote(str(invoice_id), safe='')}",
            body={"active": False},
            options=options,
        )
        return self.http.parseResponseBody(response)

    def getInstallmentPayments(self, installment_id: int, options: Optional[RequestOptions] = None) -> Any:
        response = self.http.request(
            "GET",
            f"/v1/payment-installments/{quote(str(installment_id), safe='')}/payments",
            options=options,
        )
        return self.http.parseResponseBody(response)

File: node-sdk/src/endpoints.ts
import { PviumHttpClient } from "./client";
import {
  CreateInvoiceRequest,
  RequestOptions,
  CancelInvoiceResponse,
  CreateInvoiceResponse,
  ListInvoicesResponse,
  InvoiceStatusResponse,
  InstallmentPaymentsResponse,
} from "./types";

export class PviumEndpoints {
  constructor(private readonly http: PviumHttpClient) {}

  async createInvoice(
    body: CreateInvoiceRequest,
    options?: RequestOptions,
  ): Promise<CreateInvoiceResponse> {
    const response = await this.http.request({
      method: "POST",
      path: "/v1/invoices",
      body,
      options,
    });
    return this.http.parseResponseBody<CreateInvoiceResponse>(response);
  }

  async listInvoices(options?: RequestOptions): Promise<ListInvoicesResponse> {
    const response = await this.http.request({
      method: "GET",
      path: "/v1/invoices",
      options,
    });
    return this.http.parseResponseBody<ListInvoicesResponse>(response);
  }

  async getInvoiceStatus(
    code: string,
    options?: RequestOptions,
  ): Promise<InvoiceStatusResponse> {
    const response = await this.http.request({
      method: "GET",
      path: `/v1/invoices/${encodeURIComponent(code)}/status`,
      options,
    });

    return this.http.parseResponseBody<InvoiceStatusResponse>(response);
  }

  async cancelInvoice(
    id: string | number,
    options?: RequestOptions,
  ): Promise<CancelInvoiceResponse> {
    const response = await this.http.request({
      method: "PATCH",
      path: `/v1/invoices/${encodeURIComponent(String(id))}`,
      body: { active: false },
      options,
    });

    return this.http.parseResponseBody<CancelInvoiceResponse>(response);
  }

  async getInstallmentPayments(
    id: number,
    options?: RequestOptions,
  ): Promise<InstallmentPaymentsResponse> {
    const response = await this.http.request({
      method: "GET",
      path: `/v1/payment-installments/${encodeURIComponent(String(id))}/payments`,
      options,
    });

    return this.http.parseResponseBody<InstallmentPaymentsResponse>(response);
  }
}
